File: form/wicket/src/main/java/org/opensingular/form/wicket/mapper/maps/MarkableGoogleMapsPanel.java
/*
 * Copyright (C) 2016 Singular Studios (a.k.a Atom Tecnologia) - www.opensingular.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.opensingular.form.wicket.mapper.maps;

import static org.opensingular.lib.wicket.util.util.Shortcuts.*;

import com.google.maps.internal.PolylineEncoding;
import com.google.maps.model.LatLng;
import org.apache.commons.collections.CollectionUtils;
import org.apache.commons.lang3.StringEscapeUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.wicket.Component;
import org.apache.wicket.ajax.AjaxRequestTarget;
import org.apache.wicket.markup.ComponentTag;
import org.apache.wicket.markup.head.IHeaderResponse;
import org.apache.wicket.markup.head.JavaScriptReferenceHeaderItem;
import org.apache.wicket.markup.head.OnDomReadyHeaderItem;
import org.apache.wicket.markup.html.WebComponent;
import org.apache.wicket.markup.html.WebMarkupContainer;
import org.apache.wicket.markup.html.basic.Label;
import org.apache.wicket.markup.html.form.Button;
import org.apache.wicket.markup.html.form.FormComponent;
import org.apache.wicket.markup.html.form.HiddenField;
import org.apache.wicket.markup.html.link.ExternalLink;
import org.apache.wicket.model.IModel;
import org.apache.wicket.model.LoadableDetachableModel;
import org.apache.wicket.model.Model;
import org.apache.wicket.request.resource.PackageResourceReference;
import org.json.JSONObject;
import org.opensingular.form.SInstance;
import org.opensingular.form.type.util.SILatitudeLongitudeList;
import org.opensingular.form.type.util.STypeLatitudeLongitude;
import org.opensingular.form.type.util.STypeLatitudeLongitudeGMaps;
import org.opensingular.form.view.SViewCurrentLocation;
import org.opensingular.form.wicket.model.SInstanceFieldModel;
import org.opensingular.form.wicket.model.SInstanceValueModel;
import org.opensingular.lib.commons.base.SingularProperties;
import org.opensingular.lib.commons.lambda.ISupplier;
import org.opensingular.lib.wicket.util.bootstrap.layout.BSContainer;
import org.opensingular.lib.wicket.util.bootstrap.layout.TemplatePanel;
import org.opensingular.lib.wicket.util.util.WicketUtils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static org.opensingular.lib.wicket.util.util.Shortcuts.*;
public class MarkableGoogleMapsPanel<T> extends BSContainer<MarkableGoogleMapsPanel<T>> {


    private static final String PANEL_SCRIPT  = "MarkableGoogleMapsPanel.js";

    private static final String                             SINGULAR_GOOGLEMAPS_JS_KEY     = "singular.googlemaps.js.key";
    private static final String                             SINGULAR_GOOGLEMAPS_STATIC_KEY = "singular.googlemaps.static.key";
    public static final String                              MAP_ID                         = "map";
    public static final String                              MAP_STATIC_ID                  = "mapStatic";
    private final LatLongMarkupIds                          ids;

    private final String                                    singularKeyMaps                = SingularProperties.getOpt(SINGULAR_GOOGLEMAPS_JS_KEY).orElse(null);
    private final String                                    singularKeyMapStatic           = SingularProperties.getOpt(SINGULAR_GOOGLEMAPS_STATIC_KEY).orElse(null);

    private final IModel<String>                            metaDataModel                  = new Model<>();
    private final boolean                                   visualization;
    private final ISupplier<? extends SViewCurrentLocation> viewSupplier;


    private boolean           multipleMarkers;
    private String            callbackUrl;
    private String            tableContainerId;

    private final Button       clearButton;
    private final Button       currentLocationButton;
    private final WebMarkupContainer verNoMaps;
    private final ImgMap       mapStatic;
    private final WebMarkupContainer  map      = new WebMarkupContainer(MAP_ID);
    private final HiddenField<String> metaData = new HiddenField<>("metadados", metaDataModel);


    @Override
    public void renderHead(IHeaderResponse response) {
        super.renderHead(response);
        final PackageResourceReference customJS = new PackageResourceReference(getClass(), PANEL_SCRIPT);
        response.render(JavaScriptReferenceHeaderItem.forReference(customJS));
        if (StringUtils.isNotBlank(singularKeyMapStatic) && StringUtils.isNotBlank(singularKeyMaps)) {
            response.render(OnDomReadyHeaderItem.forScript("Singular.createSingularMap(" + stringfyId(metaData) + ", '" + singularKeyMaps + "');"));
        }
    }

    public void updateJS(AjaxRequestTarget target) {
        if (StringUtils.isNotBlank(singularKeyMapStatic) && StringUtils.isNotBlank(singularKeyMaps)) {
            target.appendJavaScript("window.setTimeout(function () {Singular.createSingularMap(" + stringfyId(metaData) + ", '" + singularKeyMaps + "');}, 500);");
        }
    }

    public MarkableGoogleMapsPanel(LatLongMarkupIds ids, IModel<? extends SInstance> model, ISupplier<? extends SViewCurrentLocation> viewSupplier, boolean visualization, boolean multipleMarkers) {
        super(model.getObject().getName());
        this.visualization = visualization;
        this.ids = ids;
        this.viewSupplier = viewSupplier;
        this.clearButton = new Button("clearButton", $m.ofValue("Limpar"));
        this.currentLocationButton = new Button("currentLocationButton", $m.ofValue("Marcar Minha Posição"));
        this.multipleMarkers = multipleMarkers;

        IModel<SInstance> zoomModel = new SInstanceValueModel<>(new SInstanceFieldModel<>(model, STypeLatitudeLongitudeGMaps.FIELD_ZOOM));

        if (!multipleMarkers) {

            IModel<SInstance> latitudeModel  = new SInstanceValueModel<>(new SInstanceFieldModel<>(model, STypeLatitudeLongitude.FIELD_LATITUDE));
            IModel<SInstance> longitudeModel = new SInstanceValueModel<>(new SInstanceFieldModel<>(model, STypeLatitudeLongitude.FIELD_LONGITUDE));
            LoadableDetachableModel<String> googleMapsLinkModel = $m.loadable(() -> {
                if (latitudeModel.getObject() != null && longitudeModel.getObject() != null) {
                    String localization = latitudeModel.getObject() + "," + longitudeModel.getObject() + "/@" + latitudeModel.getObject() + "," + longitudeModel.getObject();
                    return "https://www.google.com.br/maps/place/" + localization + "," + zoomModel.getObject() + "z";
                } else {
                    return "https://www.google.com.br/maps/search/-15.7481632,-47.8872134,15";
                }
            });

            verNoMaps = new ExternalLink("verNoMaps", googleMapsLinkModel, $m.ofValue("Visualizar no Google Maps")) {
                @Override
                protected void onComponentTag(ComponentTag tag) {
                    super.onComponentTag(tag);
                    tag.put("target", "_blank");
                }
            };
            mapStatic = new ImgMap(MAP_STATIC_ID, $m.loadable(() -> {
                String latLng = generateLatLngMaker(latitudeModel, longitudeModel);
                return configureMapToView(Arrays.asList(latLng), zoomModel, null);
            }));
        } else {
            verNoMaps = new WebMarkupContainer("verNoMaps");
            mapStatic = new ImgMap(MAP_STATIC_ID, $m.loadable(() -> {
                List<String> makerList  = new ArrayList<>();
                List<LatLng> latLngList = new ArrayList<>();
                ((SILatitudeLongitudeList) model.getObject()).getPoints().forEach(m -> {
                    IModel<?> latitude  = Model.of(m.getLatitude());
                    IModel<?> longitude = Model.of(m.getLongitude());

                    LatLng latLng = new LatLng(m.getLatitude().doubleValue(), m.getLongitude().doubleValue());
                    latLngList.add(latLng);
                    String latLngText = generateLatLngMaker(latitude, longitude);
                    makerList.add(latLngText);
                });

                return configureMapToView(makerList, zoomModel, latLngList);

            }));
        }

        clearButton.setDefaultFormProcessing(false);
        currentLocationButton.setDefaultFormProcessing(false);

    }

    private String generateLatLngMaker(IModel<?> latitudeModel, IModel<?> longitudeModel) {
        if (latitudeModel.getObject() != null && longitudeModel.getObject() != null) {
            return latitudeModel.getObject() + "," + longitudeModel.getObject();
        }
        return null;
    }

    private String configureMapToView(List<String> latLngMakerList, IModel<?> zoomModel, List<LatLng> latLngList) {
        String        latLng = "-15.7922, -47.4609";
        StringBuilder marker = new StringBuilder();
        if (CollectionUtils.isNotEmpty(latLngMakerList)) {
            latLngMakerList.stream()
                    .filter(l -> l != null && !l.isEmpty())
                    .forEach(l -> marker.append("&markers=").append(l));
            if (StringUtils.isNotEmpty(latLngMakerList.get(0))) {
                latLng = latLngMakerList.get(0);
            }
        }

        latLng = getMeanOfPoints(latLngList, latLng);

        StringBuilder parameters = new StringBuilder();
        parameters.append("key=").append(singularKeyMapStatic);
        parameters.append("&size=1000x").append(getHeight() - 35);


        parameters.append("&center=").append(latLng);
        if (multipleMarkers) {
            parameters.append("&path=color:0x0ea001AA|weight:0|fillcolor:0xFFB6C1BB");
            parameters.append("|enc:").append(PolylineEncoding.encode(latLngList));
        } else {
            parameters.append("&zoom=").append(zoomModel.getObject());
        }
        parameters.append(marker);
        return "https://maps.googleapis.com/maps/api/staticmap?" + parameters.toString();
    }

    private String getMeanOfPoints(List<LatLng> latLngList, String latLng) {
        String mean = latLng;
        if (CollectionUtils.isNotEmpty(latLngList)) {
            Model<Double> latModel = new Model<>(Double.valueOf(0));
            Model<Double> lngModel = new Model<>(Double.valueOf(0));
            latLngList.forEach(l -> {
                latModel.setObject(latModel.getObject() + l.lat);
                lngModel.setObject(lngModel.getObject() + l.lng);
            });
            String lat = String.valueOf(latModel.getObject() / latLngList.size());
            String lng = String.valueOf(lngModel.getObject() / latLngList.size());
            mean = lat + ", " + lng;
        }
        return mean;
    }

    private void populateMetaData() {
        JSONObject json = new JSONObject();
        json.put("idClearButton", clearButton.getMarkupId(true));
        json.put("idCurrentLocationButton", currentLocationButton.getMarkupId(true));
        json.put("idMap", map.getMarkupId(true));
        json.put("idLat", ids.latitudeId);
        json.put("idLng", ids.longitudeId);
        json.put("idZoom", ids.zoomId);
        json.put("readOnly", isReadOnly());
        json.put("tableContainerId", tableContainerId);
        json.put("callbackUrl", callbackUrl);
        json.put("multipleMarkers", multipleMarkers);
        metaDataModel.setObject(json.toString());
    }

    @Override
    protected void onInitialize() {
        super.onInitialize();

        TemplatePanel panelErrorMsg = newTemplateTag(tt -> {
            final StringBuilder templateBuilder = new StringBuilder();
            templateBuilder.append("<div> <label class=\"text-danger\" wicket:id='errorMapStatic'></label> </div>");
            templateBuilder.append("<div> <label class=\"text-danger\" wicket:id='errorMapJS'></label> </div>");
            return templateBuilder.toString();
        });

        TemplatePanel templatePanel = newTemplateTag(tt -> {
            final StringBuilder templateBuilder = new StringBuilder();
            templateBuilder.append(" <div class=\"form-group\"> ");
            templateBuilder.append("    <input type=\"button\" class=\"btn btn-default\" wicket:id=\"clearButton\"> ");
            templateBuilder.append("    <input type=\"button\" class=\"btn btn-default\" wicket:id=\"currentLocationButton\"> ");
            templateBuilder.append("    <a class=\"btn btn-default\" wicket:id=\"verNoMaps\"></a> ");
            templateBuilder.append(" </div>");
            templateBuilder.append(" <div wicket:id=\"map\" style=\"height: 90%;\"> </div> ");
            templateBuilder.append(" <input type=\"hidden\" wicket:id=\"metadados\"> ");
            templateBuilder.append("<div class=\"form-group\">");
            templateBuilder.append("    <img wicket:id=\"mapStatic\" > ");
            templateBuilder.append(" </div>");
            return templateBuilder.toString();
        });

        Component errorMapStatic = new Label("errorMapStatic", "Não foi encontrada a Key do Google Maps Static no arquivo singular.properties").setVisible(false);
        Component errorMapJS = new Label("errorMapJS", "Não foi encontrada a Key do Google Maps JS no arquivo singular.properties").setVisible(false);

        panelErrorMsg.add(errorMapJS, errorMapStatic);
        templatePanel.add(verNoMaps, clearButton, currentLocationButton, map, metaData, mapStatic);

        if (StringUtils.isBlank(singularKeyMapStatic)) {
            templatePanel.setVisible(false);
            errorMapStatic.setVisible(true);
        }
        if (StringUtils.isBlank(singularKeyMaps)) {
            templatePanel.setVisible(false);
            errorMapJS.setVisible(true);
        }
    }

    @Override
    protected void onConfigure() {
        super.onConfigure();

        visitChildren(FormComponent.class, (comp, visit) -> comp.setEnabled(!isVisualization()));
        this.add(WicketUtils.$b.attrAppender("style", "height: " + getHeight() + "px;", ""));

        map.setVisible(!isVisualization());
        currentLocationButton.setVisible(SViewCurrentLocation.class.isInstance(viewSupplier.get()) && !isVisualization());

        boolean notMultipleAndNotVisualization = !multipleMarkers && !isVisualization();
        clearButton.setVisible(notMultipleAndNotVisualization);
        mapStatic.setVisible(isVisualization());
        verNoMaps.setVisible(!multipleMarkers && isVisualization());
    }

    @Override
    protected void onBeforeRender() {
        super.onBeforeRender();
        populateMetaData();
    }

    protected Integer getHeight() {
        return 500;
    }

    private String stringfyId(Component c) {
        return "'" + c.getMarkupId(true) + "'";
    }

    private boolean isVisualization() {
        return visualization;
    }

    private boolean isReadOnly() {
        SViewCurrentLocation view = viewSupplier.get();
        if (view != null)
            return visualization || view.isDisableUserLocationSelection();
        return visualization;
    }

    public void enableMultipleMarkers(String callbackUrl, String tableContainerId) {
        this.multipleMarkers = true;
        this.callbackUrl = callbackUrl;
        this.tableContainerId = tableContainerId;
    }

    private static class ImgMap extends WebComponent {
        ImgMap(String id, IModel<?> model) {
            super(id, model);
        }

        @Override
        protected void onComponentTag(ComponentTag tag) {
            super.onComponentTag(tag);
            checkComponentTag(tag, "img");
            tag.put("src", StringEscapeUtils.unescapeHtml4(getDefaultModelObjectAsString()));
        }
    }

}

File: form/wicket/src/main/java/org/opensingular/form/wicket/mapper/maps/MarkableGoogleMapsPanel.js
/*
 * Copyright (C) 2016 Singular Studios (a.k.a Atom Tecnologia) - www.opensingular.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
(function () {
    "use strict";

    if (window.Singular === undefined) {
        window.Singular = function () {
        };
    }

    if (window.SingularMaps === undefined) {
        window.SingularMaps = [];
    }

    window.Singular.createSingularMap = function (idMetadados, googleMapsKey) {

        if (typeof google !== 'undefined') {
            createSingularMapImpl(idMetadados);
        } else {
            //carregar o script do gmaps apenas uma vez e recriar os demais mapas
            if (window.SingularMaps.length === 0) {
                $.getScript('https://maps.googleapis.com/maps/api/js?key=' + googleMapsKey + '&language=pt-BR')
                    .done(function () {
                        for (var i = 0; i < window.SingularMaps.length; i++) {
                            window.SingularMaps[i]();
                        }
                    })
                    .fail(function (s, Status) {
                        if (Status === 'OverQuotaMapError') {
                            var meta = JSON.parse(document.getElementById(idMetadados).value);
                            document.getElementById(meta.idMap).style.visibility = "hidden";
                        }
                    });
            }
            window.SingularMaps.push(function () {
                Singular.createSingularMap(idMetadados);
            });
        }
    };


    function createSingularMapImpl(idMetadados) {
        var metadados = JSON.parse(document.getElementById(idMetadados).value);
        var zoomElement = document.getElementById(metadados.idZoom);
        if (document.getElementById(metadados.idLat) !== null) {
            var latElement = document.getElementById(metadados.idLat);
            var lngElement = document.getElementById(metadados.idLng);
            configureMap(latElement, lngElement, zoomElement, metadados.idMap, metadados.idClearButton, JSON.parse(metadados.readOnly), metadados.idCurrentLocationButton);
        } else if (metadados.multipleMarkers) {
            var tableContainerElement = document.getElementById(metadados.tableContainerId);
            var urlKml = metadados.urlKml;
            configureMapMultipleMarkers(tableContainerElement, zoomElement, metadados.idMap, metadados.idClearButton, JSON.parse(metadados.readOnly), metadados.callbackUrl, urlKml);
        } else {
            document.getElementById(metadados.idMap).style.visibility = "hidden";
        }
    }

    function configureMap(latElement, lngElement, zoomElement, idMap, idClearButton, readOnly, idCurrentLocationButton) {
        var latLong = buildGmapsLatLong(latElement.value, lngElement.value);

        var map = new google.maps.Map(document.getElementById(idMap), {
            zoom: Number(zoomElement.value),
            center: latLong
        });

        map.addListener('zoom_changed', function () {
            zoomElement.value = map.zoom;
        });

        var marker = configureMarker(latLong, latElement, lngElement, map, readOnly);
        configureFieldsEvents(latElement, lngElement, map, marker, idClearButton, readOnly, idCurrentLocationButton);

        if (!readOnly) {
            map.addListener('click', function (event) {
                if (!marker.getVisible()) {
                    marker.setVisible(true);
                }
                marker.setPosition(event.latLng);
                latElement.value = event.latLng.lat();
                lngElement.value = event.latLng.lng();
            });
        }
        return map;
    }

    function configureMapMultipleMarkers(tableContainerElement, zoomElement, idMap, idClearButton, readOnly, callbackUrl, urlKml) {
        var markers = [];
        var latLong = buildGmapsLatLong();
        var polygon = new google.maps.Polygon({
            strokeColor: '#FF0000',
            strokeOpacity: 0.8,
            strokeWeight: 2,
            fillColor: '#FF0000',
            fillOpacity: 0.35
        });

        var map = new google.maps.Map(document.getElementById(idMap), {
            zoom: Number(zoomElement.value),
            center: latLong
        });

        //O problema não é o size da url.

       // urlKml = "http://showcase.opensingular.org/download/Flwy37pGExSWP38Vm08zdLphfQV6c9lo79JXTwp9/97853f29-0270-4902-8817-b7d90e1b5fc9";
        // urlKml = "http://googlemaps.github.io/js-v2-samples/ggeoxml/cta.kml";
        //   urlKml = "https://sites.google.com/a/gmapas.com/home/poligonos-ibge/poligonos-municipios-ibge-rio-grande-do-norte/Municipios_RN.kml"
      //  urlKml = "http://api.flickr.com/services/feeds/geo/?g=322338@N20&lang=en-us&format=feed-georss";
        console.log(urlKml);
        if(urlKml !== '' && urlKml != null){
            new google.maps.KmlLayer({
                url: urlKml,
                map: map
            });
        } else {
            if (!readOnly) {
                map.addListener('zoom_changed', function () {
                    zoomElement.value = map.zoom;
                });
            }

            configureMarkers(tableContainerElement, map, readOnly, markers, polygon);
            draw(map, polygon, markers);
            if (!readOnly) {
                configureMultipleFieldsEvents(tableContainerElement, map, markers, polygon);
            }

            if (!readOnly) {
                map.addListener('click', function (event) {
                    var params = {'lat': event.latLng.lat(), 'lng': event.latLng.lng()};
                    Wicket.Ajax.post({u: callbackUrl, ep: params});

                    var number = countMarkers(tableContainerElement);
                    var marker = createMarker(map, event.latLng, polygon, readOnly, true, number + 1);
                    markers.push(marker);
                    draw(map, polygon, markers);
                });
            }
        }
        return map;
    }

    function configureMarkers(tableContainerElement, map, readOnly, markers, polygon) {
        jQuery(tableContainerElement)
            .find('.list-table-body table tbody tr')
            .each(function (index) {
                var valLat;
                var valLng;
                if (readOnly) {
                    var latLongElements = $(this).find('td');
                    valLat = $.trim($(latLongElements[0]).text());
                    valLng = $.trim($(latLongElements[1]).text());
                } else {
                    var latLongElements = $(this).find('input[type=text]');
                    valLat = latLongElements[0].value;
                    valLng = latLongElements[1].value;
                }
                var latLng;
                if (isLatLongNotEmpty(valLat, valLng)) {
                    latLng = buildGmapsLatLong(valLat, valLng);
                }

                markers.push(createMarker(map, latLng, polygon, readOnly, false, index+1));
            })
        ;
    }

    function createMarker(map, latLng, polygon, readOnly, animate, number) {
        var marker = new google.maps.Marker({
            map: map,
            visible: false,
            label: number.toString()
        });

        if (latLng) {
            marker.setPosition(latLng);
            marker.setVisible(true);
        }

        if (animate) {
            marker.setAnimation(google.maps.Animation.DROP);
        }

        return marker;
    }

    function countMarkers(tableContainerElement) {
        return jQuery(tableContainerElement)
            .find('.list-table-body table tbody tr')
            .length;
    }

    function findLatLongRow(tableContainerElement, markers, marker) {
        var i = 0;

        while (i < markers.length) {
            if (markers[i].getPosition().equals(marker.getPosition())) {
                break;
            }
            i++;
        }

        var line = jQuery(tableContainerElement)
            .find('.list-table-body table tbody tr')[i];
        return jQuery(line).find('input[type=text]');
    }

    function draw(map, polygon, markers) {
        var visibleMarkers = markers.filter(function (m) { return m.getVisible();});
        if (visibleMarkers.length > 2) {
            var coords = visibleMarkers.map(function (m) {
                return {lat: m.getPosition().lat(), lng: m.getPosition().lng()};
            });

            polygon.setPaths(coords);
            polygon.setMap(map);

        } else {
            polygon.setMap(null);
        }
    }

    function configureMultipleFieldsEvents(tableContainerElement, map, markers, polygon) {

        jQuery(tableContainerElement)
            .find('.list-table-body table tbody tr')
            .each(function (index) {
                var latLongElements = $(this).find('input[type=text]');
                var latElement = latLongElements[0];
                var lngElement = latLongElements[1];
                var marker = markers[index];
                var center = markers.length === 1;

                $(latElement).on('change', function () {
                    defineMarkerPositionManual(latElement, lngElement, map, marker, center);
                    draw(map, polygon,  markers);
                });
                $(lngElement).on('change', function () {
                    defineMarkerPositionManual(latElement, lngElement, map, marker, center);
                    draw(map, polygon,  markers);
                });

            })
        ;

    }

    function configureMarker(latLong, latElement, lngElement, map, readOnly) {
        var marker = new google.maps.Marker({
            position: latLong,
            map: map
        });

        if (isLatLongNotEmpty(latElement.value, lngElement.value)) {
            marker.setPosition(latLong);
            marker.setVisible(true);
        } else {
            marker.setVisible(false);
        }

        marker.addListener('click', function () {
            if (marker.getVisible()) {
                marker.setVisible(false);
                latElement.value = null;
                lngElement.value = null;
            }
        });

        return marker;
    }

    function buildGmapsLatLong(lat, lng) {
        var valLat;
        var valLng;
        if (!lat && !lng) {
            valLat = -15.7922;
            valLng = -47.4609;
        } else {
            valLat = lat.replace(",", ".");
            valLng = lng.replace(",", ".");
        }
        return new google.maps.LatLng(valLat, valLng);
    }

    function configureFieldsEvents(latElement, lngElement, map, marker, idClearButton, readOnly, idCurrentLocationButton) {
        if (readOnly) {
            $(latElement).attr('readonly', 'readonly');
            $(latElement).addClass('disabled');
            $(lngElement).attr('readonly', 'readonly');
            $(lngElement).addClass('disabled');
        }
        $(latElement).on('change', function () {
            defineMarkerPositionManual(latElement, lngElement, map, marker, true);
        });
        $(lngElement).on('change', function () {
            defineMarkerPositionManual(latElement, lngElement, map, marker, true);
        });
        $("#" + idClearButton).on('click', function () {
            latElement.value = null;
            lngElement.value = null;
            marker.setVisible(false);
        });
        $("#" + idCurrentLocationButton).on('click', function () {
            findCurrentLocation(
                function (success) {
                    $(latElement).val(success.coords.latitude);
                    $(lngElement).val(success.coords.longitude);
                    $(latElement).trigger('change');
                    $(lngElement).trigger('change');
                    map.setZoom(17);
                    marker.setVisible(true);
                },
                function (failure) {
                    toastr.error(failure.message);
                }
            );
        });
    }

    function isLatLongNotEmpty(valLat, valLng) {
        return valLat !== null && valLng !== "" && valLat && valLng !== "";
    }

    function defineMarkerPositionManual(latElement, lngElement, map, marker, center) {
        if (isLatLongNotEmpty(latElement.value, lngElement.value)) {
            var latLong = buildGmapsLatLong(latElement.value, lngElement.value);
            marker.setPosition(latLong);
            if (!marker.getVisible()) {
                marker.setVisible(true);
            }
            if (center) {
                map.setCenter(latLong);
            }
        } else {
            marker.setVisible(false);
        }
    }

    function findCurrentLocation(success, failure) {
        if (navigator.geolocation) {
            var options = {
                enableHighAccuracy: true,
                timeout: 5000,
                maximumAge: 0
            };
            navigator.geolocation.getCurrentPosition(success, failure, options);
        } else {
            toastr.error("Seu navegador não suporta geolocalização.");
        }

    }

})();